src/LinuxCaptureService/LinuxCaptureService.cpp

```diff
@@ -147,6 +149,90 @@ class ProducerEventProcessorHijackingFunctionEntryExitForLinuxTracing
+
+    memory_watchdog_thread = std::thread{
+        [watchdog_threshold_bytes, stop_capture_mutex, stop_capture, stop_capture_reason] {
+          while (true) {
+            {
+              absl::MutexLock lock{&*stop_capture_mutex};
+              if (stop_capture_mutex->AwaitWithTimeout(absl::Condition(&*stop_capture),
+                                                       kWatchdogPollInterval)) {
+                LOG("Stopping memory watchdog as the capture was stopped");
+                break;
+              }
+            }
+
+            // Repeatedly poll `/proc/[pid]/stat` for the resident set size (rss) of the current
+            // process.
+            std::optional<uint64_t> rss_bytes = ReadRssInBytesFromProcPidStat();
+            if (!rss_bytes.has_value()) {
+              ERROR_ONCE("Reading resident set size of the current process");
+              continue;
+            }
+            if (rss_bytes.value() > watchdog_threshold_bytes) {
+              LOG("Memory threshold exceeded: stopping capture (and stopping memory watchdog)");
+              absl::MutexLock lock{&*stop_capture_mutex};
+              *stop_capture = true;
+              *stop_capture_reason = StopCaptureReason::kMemoryWatchdog;
+              break;
+            }
+          }
+        }};
+  }
+
+  // ServerReaderWriter::Read blocks until the client has called WritesDone, or until we finish the
+  // gRPC (before the client has called WritesDone). In the latter case, the Read unblocks *after*
+  // LinuxCaptureService::Capture has returned, so we need to keep the thread around and join it at
+  // a later time (we don't want to just detach it).
+  wait_for_stop_capture_request_thread_ =
+      std::thread{[reader_writer, stop_capture_mutex, stop_capture, stop_capture_reason] {
+        orbit_grpc_protos::CaptureRequest request;
+        // The client asks for the capture to be stopped by calling WritesDone. At that point, this
+        // call to Read will return false. In the meantime, it blocks if no message is received.
+        // Read also unblocks and returns false if the gRPC finishes.
+        while (reader_writer->Read(&request)) {
+        }
+
+        absl::MutexLock lock{&*stop_capture_mutex};
+        if (!*stop_capture) {
+          LOG("Client finished writing on Capture's gRPC stream: stopping capture");
+          *stop_capture = true;
+          *stop_capture_reason = StopCaptureReason::kClientStop;
+        } else {
+          LOG("Client finished writing on Capture's gRPC stream or the RPC has already finished; "
+              "the capture was already stopped");
+        }
+      }};
+
+  stop_capture_mutex->LockWhen(absl::Condition(&*stop_capture));
+  CHECK(stop_capture_reason->has_value());
+  StopCaptureReason stop_capture_reason_to_return = stop_capture_reason->value();
+  stop_capture_mutex->Unlock();
+
+  if (memory_watchdog_thread.joinable()) {
+    memory_watchdog_thread.join();
+  }
```

Comment:
If I get this right we are blocking the thread that called this whole function until the `memory_watchdog_thread` has finished, right? Why not execute the code from `memory_watchdog_thread` here instead synchronously?

Reply:
I think you're right, this was different in the first commit because the watchdog wasn't necessarily started. I'll look into this now.

Reply:
Done, I had to give it a round of testing.